### PR TITLE
Unify argot binaries

### LINKED ISSUE #87: Unified CLI tool
Right now argot consists of multiple binaries such as `taint`, `backtrace`, `dependencies` etc. This was a historical decision made when the tools were more independent, but now they share more code.

Recently, there seems to be a trend of having a single binary that has sub-commands for various tools. An example of this is the new `nix` [command](https://nixos.wiki/wiki/Nix_command) which replaces the old `nix-*` commands.

We can do a similar thing for argot where a user invokes a single `argot` binary and specifies a subcommand for the particular tool they want to use. For example, `argot taint -config ...` instead of the old `taint -config ...`.

### PR DESCRIPTION
**Pull request by samarth-aws**
*Issue #, if available:*
#87 

*Description of changes:*
Instead of having a binary for each tool, we now have a single `argot`
binary that has each tool as a sub-command: e.g., `taint -config ...` =>
`argot taint -config ...`

This is a breaking change!

- Moves all binaries into a single cmd/argot/ folder
  - Each tool has its own folder (e.g., cmd/taint => cmd/argot/taint)
  - cmd/main.go has the single main() function
- Keeps racerg as a separate binary since it has non-Go dependencies
- Adds and changes some flags

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -32,14 +38,12 @@ shared options.
...
-The executables are in the `cmd` folder, we have one executable per tool listed.
+The executables are in the `cmd` folder. There are currently only two: `argot` and `racerg` (an experimental static data race detector).
 
 There is user documentation in the `doc` folder.
 
-The test data is in the `testdata` folder. All the Go source files used in the tests are in `testdata/src`.
-
 The library code, and most of the analysis implementations, is in the `analysis` folder. The main entry points are in
-the `load_progam.go` file for loading program and `analyzers.go` to call analyzers. The rest is organized in subfolders:
+the `load_progam.go` file for loading the program and `analyzers.go` to call analyzers. The rest is organized in subfolders:
```
Comment (victornicolet) [inline README.md]: There's a typo here (not introduced by your changes) in `load_progam.go`

File: Makefile [hunk 2]
```diff
@@ -8,9 +8,9 @@
 # Install ineffassign with: go install github.com/gordonklaus/ineffassign@latest
 # Install golint with:      go install golang.org/x/lint/golint@latest
 
-all: lint maypanic statistics reachability dependencies static-commands render taint compare defer packagescan backtrace argot-cli racerg setup-precommit
+all: setup-precommit lint racerg argot-build test
```
Comment (victornicolet) [inline Makefile]: nit: I'd like to just have to type `make argot` or `make build` instead of `make argot-build`

File: cmd/argot/main.go [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+  - render: renders a graph representation of the callgraph, or prints the program's SSA form
+  - ssa-statistics: prints statistics about the SSA representation of the program
+Examples:
+  Run the interactive CLI: argot cli --config=config.yaml main.go
+  Run the taint analysis: argot taint --config=config.yaml main.go`
+
+//gocyclo:ignore
+func main() {
+	if len(os.Args) < 2 {
+		fmt.Fprintf(os.Stderr, "error: expected subcommand\n%s", usage)
```
Comment (victornicolet) [inline cmd/argot/main.go]: Missing a new line after the usage message: 
Screenshot 2024-09-12 at 9 13 19 AM

File: doc/08_packagescan.md [hunk 4]
```diff
@@ -4,15 +4,16 @@ packagescan is a tool that searches the codebase of a program and identifies eac
...
 
 The -rawfile option is not yet fully implemented and will dump each observation of a public symbol from the target package(s) to a file.  So for example instead of just seeing which packages import the S3 service, you could see precisely which S3 APIs are called from within the program and from what functions.
 
 The example output below illustrates running packagescan on itself with the default to report on any packages reachable from the code that import "unsafe" and thus might warrant greater scrutiny.  The differences between Windows and Unix are dramatic, but there are even subtle differences between Unix variants.   Note that the table repeats the platform name rather than using column headers and "X" to indicate inclusion because column headers tend to scroll offscreen.  
 ````
-bin/packagescan ./cmd/packagescan
+$ argot packagescan ./cmd/packagescan
```
Comment (victornicolet) [inline doc/08_packagescan.md]: Should be `argot packagescan ./cmd/argot` for example

File: README.md [hunk 5]
```diff
@@ -6,24 +6,30 @@
 
 ## Overview
 
-Argot is a collection of tools:
-- `taint` performs a taint analysis on a given program.
+Argot is a collection of static analysis tools:
+- `taint` performs a taint analysis on a given program
 - `backtrace` identifies backwards data-flow traces from function calls
-- `argot-cli` is a terminal-like interface for various part of the analysis (in `cmd/cli`)
+- `argot-cli` is a terminal-like interface for parts of the analysis (in `cmd/cli`)
```
Comment (victornicolet) [inline README.md]: Should this be `cli`?

File: cmd/argot/cli/analysis.go [hunk 6]
```diff
@@ -350,7 +350,7 @@ func cmdTaint(tt *term.Terminal, c *dataflow.AnalyzerState, _ Command) bool {
 }
 
 // cmdBacktrace runs the backtrace analysis.
-func cmdBacktrace(tt *term.Terminal, c *dataflow.AnalyzerState, _ Command) bool {
+func cmdBacktrace(tt *term.Terminal, c *dataflow.AnalyzerState, _ Command, _ bool) bool {
 	if c == nil {
 		writeFmt(tt, "\t- %s%s%s: run the backtrace analysis with parameters in config.\n",
 			tt.Escape.Blue, cmdTaintName, tt.Escape.Reset)
```
Comment (victornicolet) [inline cmd/argot/cli/analysis.go]: Looks like a mistake in previous changes, this should be `cmdBacktraceName`

**Review by victornicolet — COMMENTED** [review]
I think there's some places that also need updating in the doc:
- l 483 in doc/01_taint.md the binary name should be replaced (and you can update the test data location too, which should be in analysis/escape)

Also you'll probably need to rebase on mainline for the tests to pass.

File: README.md (hunk 1 above)
Comment (samarth-aws) [reply README.md]: Where is the typo?

Comment (victornicolet) [reply README.md]: It should be `load_program.go` (missing `r`)

**Review by victornicolet — APPROVED** [review]

**Review by amzn-jasonrk — APPROVED** [review]

**Pull request merged**